packages/core/src/requestforattestation/RequestForAttestation.ts

```diff
@@ -34,316 +108,356 @@ function makeSigningData(
+ * the signature over the presentation **must** be generated with the full DID in order for the verification to be successful.
+ * On the other hand, a light DID that has been migrated and then deleted from the chain will not be allowed to generate valid presentations anymore.
+ *
+ * @param input - The [[RequestForAttestation]].
+ * @param verificationOpts Additional options to retrieve the details from the identifiers inside the request for attestation.
+ * @param verificationOpts.resolver - The resolver used to resolve the claimer's identity. Defaults to [[DidResolver]].
+ * @param verificationOpts.challenge - The expected value of the challenge. Verification will fail in case of a mismatch.
+ * @throws [[ERROR_IDENTITY_MISMATCH]] if the DidDetails do not match the claim owner or if the light DID is used after it has been upgraded.
+ * @returns Whether the signature is correct.
+ */
+export async function verifySignature(
+  input: IRequestForAttestation,
+  {
+    challenge,
+    resolver = DidResolver,
+  }: {
+    challenge?: string
+    resolver?: IDidResolver
+  } = {}
+): Promise<boolean> {
+  const { claimerSignature } = input
+  if (!claimerSignature) return false
+  if (challenge && challenge !== claimerSignature.challenge) return false
+  const signingData = makeSigningData(input, claimerSignature.challenge)
+  const { verified } = await DidUtils.verifyDidSignature({
+    signature: claimerSignature,
+    message: signingData,
+    expectedVerificationMethod: KeyRelationship.authentication,
+    resolver,
+  })
+  return verified
+}
 
-    // check legitimations
-    Credential.validateLegitimations(input.legitimations)
+export type Options = {
+  legitimations?: ICredential[]
+  delegationId?: IDelegationNode['id']
+}
+
+/**
+ * Builds a new [[RequestForAttestation]] object, from a complete set of required parameters.
+ *
+ * @param claim An `IClaim` object the request for attestation is built for.
+ * @param option Container for different options that can be passed to this method.
+ * @param option.legitimations Array of [[Credential]] objects of the Attester which the Claimer requests to include into the attestation as legitimations.
+ * @param option.delegationId The id of the DelegationNode of the Attester, which should be used in the attestation.
+ * @returns A new [[RequestForAttestation]] object.
+ */
+export function fromClaim(
+  claim: IClaim,
+  { legitimations, delegationId }: Options = {}
+): IRequestForAttestation {
+  const { hashes: claimHashes, nonceMap: claimNonceMap } =
+    Claim.hashClaimContents(claim)
 
-    return true
+  const rootHash = calculateRootHash({
+    legitimations,
+    claimHashes,
+    delegationId,
+  })
+
+  // signature will be added afterwards!
+  const request = {
+    claim,
+    legitimations: legitimations || [],
+    claimHashes,
+    claimNonceMap,
+    rootHash,
+    delegationId: delegationId || null,
   }
+  verifyDataStructure(request)
+  verifyDataIntegrity(request)
```

Comment:
Do other from* functions verify integrity?

Reply:
I was thinking, that it might not even hurt to always do both in from functions. What do you think?